### PR TITLE
MVP for the clickhouse module

### LINKED ISSUE #48: clickhouse_role: create MVP for the module
##### SUMMARY
clickhouse_role: create MVP for the module

To accomplish this, I would follow the following algorithm:
1. Set up a dev environment using the [quick start guide](https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html)
2. Add integration tests (yes, exactly now) by copying recursively the `/test/integration/targets/clickhouse_db` directory to `/test/integration/targets/clickhouse_role`
3. Replace all the content of `./tasks/initial.yml` in that directory with something like `- meta: end_play` for now as there's nothing to test
4. Run the integration tests with `ansible-test integration clickhouse_role --docker ubuntu2204 -vvv`, they should pass
5. Copy `plugins/modules/clickhouse_db.py` to `plugins/modules/clickhouse_role.py`
6. For MVP, add only the following arguments to your module: `state: present|absent` and `name: rolename`. Update the DOCUMENTATION and EXAMPLES section correspondingly.
7. You can use the code of clickhouse_db that is already in the file. Simplify it. To learn what queries you need to execute to achieve the result, see this [page](https://clickhouse.com/docs/en/sql-reference/statements/create/role). Basically, it's only `CREATE ROLE <rolename>` (when the state is present and the role does not exist) and `DROP ROLE <rolename>` (when the state is absent and the role exists). To figure out whether the role exists or not query the [system.roles](https://clickhouse.com/docs/ru/operations/system-tables/roles) table with `SELECT 1 (or name if it doesn't work) FROM system.roles WHERE name = '<rolename>'` and check the result.
8. When it's done, run sanity tests with `ansible-tests sanity plugins/modules/clickhouse_role.py --docker`: it'll help you catch some code syntax errors besides sanity errors
9. When the sanity run is OK, replace the `- meta: end_play` with a real test in tests/integration/targets/clickhouse_role/tasks/initial.yml with a module invocation similarly as in the same file but for clickhouse_db target.
10. Run the tests. After they are green, submit a PR

If you got any questions along the way, just ask

### PR DESCRIPTION
**Pull request by oraNod**
##### SUMMARY
Resolves #48 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
MVP clickhouse role

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

File: plugins/modules/clickhouse_role.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+attributes:
+  check_mode:
+    description: Supports check_mode.
+    support: full
+
+version_added: '0.3.0'
+
+author:
+  - Andrew Klychkov (@Andersson007)
```
Comment (Andersson007) [inline plugins/modules/clickhouse_role.py:27]:
```suggestion
```

File: plugins/modules/clickhouse_role.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2024, Andrew Klychkov (@Andersson007) <andrew.a.klychkov@gmail.com>
```
Comment (Andersson007) [inline plugins/modules/clickhouse_role.py:4]:
```suggestion
# Copyright: (c) 2024, Andrew Klychkov (@Andersson007) <andrew.a.klychkov@gmail.com>
```
Replace it with yours please

File: plugins/modules/clickhouse_role.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    state: absent
+'''
+
+RETURN = r'''
+executed_statements:
+  description:
+  - Data-modifying executed statements.
+  returned: on success
+  type: list
+  sample: ['CREATE DATABASE test_db']
```
Comment (Andersson007) [inline plugins/modules/clickhouse_role.py:67]:
```suggestion
  sample: ['CREATE ROLE test_role']
```

**Review by Andersson007 — COMMENTED** [review]
@oraNod cool, thanks!
just a couple of things not to forget later

**Comment by oraNod** [comment]
@Andersson007 PTAL. I'm not sure about the integration tests, maybe they are too simple?

File: plugins/modules/clickhouse_role.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+description:
+  - Creates or removes a ClickHouse role.
+
+attributes:
+  check_mode:
+    description: Supports check_mode.
+    support: full
+
+version_added: '0.3.0'
```
Comment (Andersson007) [inline plugins/modules/clickhouse_role.py:25]:
```suggestion
version_added: '0.5.0'
```

File: plugins/modules/clickhouse_role.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    # Do the job
+    changed = False
+    role = ClickHouseRole(module, client, name)
+
+    if state == "present":
+        if not role.exists:
+            changed = role.create()
+        else:
+            pass
```
Comment (Andersson007) [inline plugins/modules/clickhouse_role.py:163]:
```suggestion
```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  check_mode: true
+  community.clickhouse.clickhouse_role:
+    state: present
+    name: test_role
+
+- name: Check the role was created
+  register: result
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:18]:
```suggestion
- name: Check the role is not present
  ansible.builtin.assert:
    that:
    - result.result[0] == []


```
Let's check here what our query above returned. May not work exactly how i put it but the idea is clear, just check the integration test output

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+####################################################################
+
+# Test
+- name: Create role in check mode
+  register: result
+  check_mode: true
+  community.clickhouse.clickhouse_role:
+    state: present
+    name: test_role
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:13]:
```suggestion
- name: Check return values
  ansible.builtin.assert:
    that:
    - result is changed
    - result.executed_statements == ['CREATE ROLE test_role']
```
in check_mode it shows the real work but doesn't change anything and we'll check it in your next test where we're querying the roles system table if it contains info about this role (i.e. if it was actually added or not and it shouldn't have been)

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
+
+# Test
+- name: Create role in real mode
+  register: result
+  community.clickhouse.clickhouse_role:
+    state: present
+    name: test_role
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:25]:
```suggestion
- name: Check return values
  ansible.builtin.assert:
    that:
    - result is changed
    - result.executed_statements == ['CREATE ROLE test_role']

```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+- name: Check the role was created
+  register: result
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:30]:
```suggestion

- name: Check the role is present
  ansible.builtin.assert:
    that:
    - result.result[0] == ['test_role']


# Test
- name: Try to create role in real mode again to check idempotency
  register: result
  community.clickhouse.clickhouse_role:
    state: present
    name: test_role
    
- name: Check return values
  ansible.builtin.assert:
    that:
    - result is not changed
    - result.executed_statements == []


```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+# Test
+- name: Remove role in check mode
+  register: result
+  check_mode: true
+  community.clickhouse.clickhouse_role:
+    state: absent
+    name: test_role
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:38]:
```suggestion

- name: Check return values
  ansible.builtin.assert:
    that:
    - result is changed
    - result.executed_statements == ['DROP ROLE test_role']
```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+- name: Check the role was removed
+  register: result
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:43]:
```suggestion

- name: Check the role is still present
  ansible.builtin.assert:
    that:
    - result.result[0] == ['test_role']
```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+- name: Remove role in real mode
+  register: result
+  community.clickhouse.clickhouse_role:
+    state: absent
+    name: test_role
+
+- name: Check the role was removed
+  register: result
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:54]:
```suggestion
    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"

- name: Check the role is absent
  ansible.builtin.assert:
    that:
    - result.result[0] == []


# Test
- name: Try to remove role in real mode again to check idempotency
  register: result
  community.clickhouse.clickhouse_role:
    state: absent
    name: test_role

- name: Check return values
  ansible.buildit.assert:
    that:
    - result is not changed
    - result.executed_statements == []
```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+# Test
+- name: Remove role in real mode
+  register: result
+  community.clickhouse.clickhouse_role:
+    state: absent
+    name: test_role
+
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:50]:
```suggestion

- name: Check return values
  ansible.builtin.assert:
    that:
    - result is changed
    - result.executed_statements == ['DROP ROLE test_role']

```

**Review by Andersson007 — COMMENTED** [review]
@oraNod thanks!
a couple of minor things  test improvements from me

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+- name: Remove role in real mode
+  register: result
+  community.clickhouse.clickhouse_role:
+    state: absent
+    name: test_role
+- name: Check return values
+  ansible.builtin.assert:
+    that:
+    - result is changed
+    - result.executed_statements == ['DROP ROLE test_role']
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:90]:
```suggestion
    - result.executed_statements == ['DROP ROLE test_role']

```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+- name: Check the role was removed
+  register: result
+  community.clickhouse.clickhouse_client:
+    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:94]:
```suggestion
    execute: "SELECT name FROM system.roles WHERE name = 'test_role'"

```

File: tests/integration/targets/clickhouse_role/tasks/initial.yml [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+  ansible.builtin.assert:
+    that:
+    - result.result[0] == []
+
+# Test
+- name: Try to remove role in real mode again to check idempotency
+  register: result
+  community.clickhouse.clickhouse_role:
+    state: absent
+    name: test_role
```
Comment (Andersson007) [inline tests/integration/targets/clickhouse_role/tasks/initial.yml:105]:
```suggestion
    name: test_role

```

**Review by Andersson007 — COMMENTED** [review]
@oraNod it's considered a good practice to have a blank line between tasks (even in redhat courses)

**Review by Andersson007 — APPROVED** [review]
@oraNod thanks for fixing all that stuff!
@aleksvagachev are you OK with merging the module? If yes, please go ahead

**Review by aleksvagachev — APPROVED** [review]

**Comment by Andersson007** [comment]
@oraNod thanks for the contribution and congratulations on getting your first module merged!:)
@aleksvagachev thanks fro reviewing and merging!

**Pull request merged**